## **Programming for Data science**

### **Project 1**
**Group members:**
* Ghita Rholi, 20221491@novaims.unl.pt
* Mariana Ferreira, 20211637@novaims.unl.pt
* Mariana Neto, 20211527@novaims.unl.pt

### Step 2. Data preparation

- **1.** Description of the dataset’s structure; 
- **2.** Features’ transformation; 
- **3.** Identification and treatment of possible data inconsistencies (like duplicates, missing values, extreme/strange/unlikely values);
    - **3.1** Duplicates
    - **3.2** Missing Values
    - **3.3** Extreme/strange/unlikely values
- **4.** Identification and treatment of inconsistent facts
- **5.** Exporting the dataset

<br>

##### Importing the necessary libraries

In [1]:
import numpy as np

import pandas as pd

<br>

##### Importing the dataset from the previous notebook

In [2]:
data = pd.read_excel("datasets_combined.xlsx", index_col = "ID")

<br>

### **1.** Description of the dataset’s structure

**1.1** By using ``.head()`` method it is possible to visualize the first 5 rows of the dataset, which allows us to recall the structure of our dataframe. 

In [3]:
data.head()

CostContact  Revenue  BirthYear   Education MaritalStatus  NKids  \
ID                                                                       
5314            3       11     1951.0  Graduation      Together    0.0   
2493            3       11     1966.0  Graduation       Married    0.0   
4931            3       11     1977.0  Graduation      Together    0.0   
1100            3       11     1960.0      Master      Together    1.0   
7224            3       11     1977.0  Graduation       Married    NaN   

      NTeens    Income  MntWines  MntFruits  ...  NWebPurchases  \
ID                                           ...                  
5314     1.0   40689.0       270          3  ...              7   
2493     1.0   61286.0       356          0  ...              7   
4931     0.0  157146.0         1          0  ...              0   
1100     2.0   41275.0        24          4  ...              3   
7224     NaN   31353.0        10          0  ...              1   

      NCatalogPurchases  NStorePurchases  NWebVisitsMonth Response1  \
ID                                                                    
5314                  1                5                8         0   
2493                  1                8                5         0   
4931                 28                0                1         0   
1100                  1                3                5         0   
7224                  1                2                8         0   

      Response2  Response3  Response4  Response5  Response  
ID                                                          
5314          0          0          0          0         0  
2493          0          0          0          0         0  
4931          0          0          0          0         0  
1100          0          0          0          0         0  
7224          0          0          0          0         0  

[5 rows x 28 columns]

<br>




**1.2**  The ``.T `` method enables the visualization of the same five rows as previously, only with all the columns being displayed.

In [4]:
data.head().T

ID                       5314        2493        4931        1100        7224
CostContact                 3           3           3           3           3
Revenue                    11          11          11          11          11
BirthYear              1951.0      1966.0      1977.0      1960.0      1977.0
Education          Graduation  Graduation  Graduation      Master  Graduation
MaritalStatus        Together     Married    Together    Together     Married
NKids                     0.0         0.0         0.0         1.0         NaN
NTeens                    1.0         1.0         0.0         2.0         NaN
Income                40689.0     61286.0    157146.0     41275.0     31353.0
MntWines                  270         356           1          24          10
MntFruits                   3           0           0           4           0
MntMeatProducts            27         107        1725          22          10
MntFishProducts            39          19           2           0           3
MntSweetProducts            6           9           1           2           1
MntGoldProds               99         117           1           9           7
CustomerFrom       2013-03-18  2013-08-02  2013-04-29  2014-03-24  2013-12-29
Complain                    0           0           0           0           0
Recency                    69          34          13          33          24
NDealsPurchases             7           2           0           4           2
NWebPurchases               7           7           0           3           1
NCatalogPurchases           1           1          28           1           1
NStorePurchases             5           8           0           3           2
NWebVisitsMonth             8           5           1           5           8
Response1                   0           0           0           0           0
Response2                   0           0           0           0           0
Response3                   0           0           0           0           0
Response4                   0           0           0           0           0
Response5                   0           0           0           0           0
Response                    0           0           0           0           0

<br>




**1.3**  With the ``.shape()`` method we can access the number of rows and columns a dataset has.

In [5]:
print("Shape of the dataframe:", data.shape)

Shape of the dataframe: (2240, 28)


The dataset is composed of 2240 rows and 28 columns.

<br>

**1.4**   The `` .info() `` method provides more information about the dataset, including the existence of missing values and the data type of the elements of each row.

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2240 entries, 5314 to 10478
Data columns (total 28 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CostContact        2240 non-null   int64  
 1   Revenue            2240 non-null   int64  
 2   BirthYear          2218 non-null   float64
 3   Education          2209 non-null   object 
 4   MaritalStatus      2240 non-null   object 
 5   NKids              2212 non-null   float64
 6   NTeens             2212 non-null   float64
 7   Income             2189 non-null   float64
 8   MntWines           2240 non-null   int64  
 9   MntFruits          2240 non-null   int64  
 10  MntMeatProducts    2240 non-null   int64  
 11  MntFishProducts    2240 non-null   int64  
 12  MntSweetProducts   2240 non-null   int64  
 13  MntGoldProds       2240 non-null   int64  
 14  CustomerFrom       2240 non-null   object 
 15  Complain           2240 non-null   int64  
 16  Recency            2

Some conclusions can be reached with this information, which are:
* The columns BirthYear, Education, NKids, NTeens and Income have missing values;
* Most of the columns have an integer data type and, therefore, numerical;
* BirthYear, NKids and NTeens are considered as floats, when they should be considered as integer, as they cannot have a decimal part (This will be solved in the features transformation stage).

<br>

**1.5**  The ``.describe()`` method provides some measurements for the numerical and categorical data of the dataframe. This allows us to get to know the numerical features and draw some conclusions. The `` include`` and ` exclude `parameters control which data types to describe, and the ``.T`` method transposes the dataframe so that we can see the measurements for all the columns. <br><br>


With data.describe() will get the default values of the ``.describe()`` method wich are the measurements of the numerical features.
Using "exclude=np.number" we can get the measurements of the categorical features.

In [7]:
display("Numerical features:", data.describe().T,
        "----------------------------------------------------------------------------------------------",
        "Non-numeric features:", data.describe(exclude=np.number))

'Numerical features:'

count          mean           std     min       25%  \
CostContact        2240.0      3.000000  0.000000e+00     3.0      3.00   
Revenue            2240.0     11.000000  0.000000e+00    11.0     11.00   
BirthYear          2218.0   1968.768711  1.198568e+01  1893.0   1959.00   
NKids              2212.0      0.444394  5.389239e-01     0.0      0.00   
NTeens             2212.0      0.507685  5.433989e-01     0.0      0.00   
Income             2189.0  97943.443582  2.136389e+06  1730.0  35340.00   
MntWines           2240.0    303.935714  3.365974e+02     0.0     23.75   
MntFruits          2240.0     26.302232  3.977343e+01     0.0      1.00   
MntMeatProducts    2240.0    166.950000  2.257154e+02     0.0     16.00   
MntFishProducts    2240.0     37.525446  5.462898e+01     0.0      3.00   
MntSweetProducts   2240.0     27.062946  4.128050e+01     0.0      1.00   
MntGoldProds       2240.0     44.021875  5.216744e+01     0.0      9.00   
Complain           2240.0      0.009375  9.639117e-02     0.0      0.00   
Recency            2240.0     49.109375  2.896245e+01     0.0     24.00   
NDealsPurchases    2240.0      2.325000  1.932238e+00     0.0      1.00   
NWebPurchases      2240.0      4.084821  2.778714e+00     0.0      2.00   
NCatalogPurchases  2240.0      2.662054  2.923101e+00     0.0      0.00   
NStorePurchases    2240.0      5.790179  3.250958e+00     0.0      3.00   
NWebVisitsMonth    2240.0      5.316518  2.426645e+00     0.0      3.00   
Response1          2240.0      0.064286  2.453160e-01     0.0      0.00   
Response2          2240.0      0.013393  1.149756e-01     0.0      0.00   
Response3          2240.0      0.072768  2.598131e-01     0.0      0.00   
Response4          2240.0      0.074554  2.627283e-01     0.0      0.00   
Response5          2240.0      0.072768  2.598131e-01     0.0      0.00   
Response           2240.0      0.149107  3.562736e-01     0.0      0.00   

                       50%       75%         max  
CostContact            3.0      3.00         3.0  
Revenue               11.0     11.00        11.0  
BirthYear           1970.0   1977.00      1996.0  
NKids                  0.0      1.00         2.0  
NTeens                 0.0      1.00         2.0  
Income             51412.0  68655.00  99999999.0  
MntWines             173.5    504.25      1493.0  
MntFruits              8.0     33.00       199.0  
MntMeatProducts       67.0    232.00      1725.0  
MntFishProducts       12.0     50.00       259.0  
MntSweetProducts       8.0     33.00       263.0  
MntGoldProds          24.0     56.00       362.0  
Complain               0.0      0.00         1.0  
Recency               49.0     74.00        99.0  
NDealsPurchases        2.0      3.00        15.0  
NWebPurchases          4.0      6.00        27.0  
NCatalogPurchases      2.0      4.00        28.0  
NStorePurchases        5.0      8.00        13.0  
NWebVisitsMonth        6.0      7.00        20.0  
Response1              0.0      0.00         1.0  
Response2              0.0      0.00         1.0  
Response3              0.0      0.00         1.0  
Response4              0.0      0.00         1.0  
Response5              0.0      0.00         1.0  
Response               0.0      0.00         1.0

'----------------------------------------------------------------------------------------------'

'Non-numeric features:'

Education MaritalStatus CustomerFrom
count         2209          2240         2240
unique           6             8          663
top     Graduation       Married   2012-08-31
freq          1107           864           12

Regarding numerical data the conclusions observed were:
* The measurement count, counts the number of rows with non-null values which also allows us to know which features have missing values. So, it provides the same information as the ``.info()`` method about missing values.
* It is clear that some features (CostContact and Revenue) have a null standard deviation which means that every value of that feature is exactly the same. Therefore, those features are not useful for this model, as they are not going to affect the target feature.
* The minimum value of BirthYear is 1893. If we consider the present date as 01/01/2018, the value mentioned represents a person with 125 years, which does not make sense. Thus, values similar to this will have to be eliminated at later stage.  
* The maximum value of the feature Income is 99999999.0, which is significantly larger than the minimum (1730.0), the mean (97943.443582) and every quartile. This implies that the maximum Income value is an outlier that is biasing the model, and should be removed, which will be done at later stages.

Regarding non-numerical data the conclusions observed were:
* It is also possible to acknowledge the existence of missing values with the measurement count, specifically in the Education column.
* It is noticeable that MaritalStatus has eight unique values, which is surprising given that the most common values for such a feature are "married" and "single". This will be explored further in a later stage.


<br>

### **2.** Features’ transformation

Some features are not being represented in the most usable format, hence certain transformations are necessary, for instance:
* As stated above, BirthYear, NKids and NTeens are stored as floating points, whereas it makes more sense to be integers. This step will be performed after the treatment of missing values, since it is not possible to convert empty spaces, i.e. NaN values or infinite values to integer.
* CustomerFrom is the date of customer’s enrollment with the company. However, CustomerFrom is stored as an object meaning that we cannot utilize it to perform useful numerical operations. Thus, it would be more usefull if that feature was replaced by the number of days since that enrollment, that is, the CustomerDays.
* BirthYear does not provide a direct value in comparison to the age. Hence, the first feature can be used to calculate a new one, to ne replaced with, the Age.

<br>

##### **2.1** Converting CustomerFrom to CustomerDays

In [8]:
# Overrides CustomerFrom after pd.to_datetime(), changing its data type from object to datetime
data["CustomerFrom"] = pd.to_datetime(data["CustomerFrom"], format="%Y/%m/%d")  
print("'CustomerFrom' data type:", data["CustomerFrom"].dtype)

'CustomerFrom' data type: datetime64[ns]


It is not enough to simply convert CustomerFrom into datetime format. In order to include it in our analysis, the feature needs to be converted into a directly measurable quantity. That is, the number of days between the date stored in the feature and some reference date.

In [9]:
# Setup reference date as the first day of 2018
reference_date = "2018/01/01"
# Computing the number of days since the reference date 
customer_days = (pd.to_datetime(reference_date, format="%Y/%m/%d") - data["CustomerFrom"]).dt.days 
# Inserting the new column (CustomerDays)
data.insert(loc=4, column="CustomerDays", value=customer_days)

# Deleting the 'BirthYear' feature, as it is being replaced by Age, that contained the same information
del data["CustomerFrom"]

# Checking the data
data.head()

CostContact  Revenue  BirthYear   Education  CustomerDays MaritalStatus  \
ID                                                                              
5314            3       11     1951.0  Graduation          1750      Together   
2493            3       11     1966.0  Graduation          1613       Married   
4931            3       11     1977.0  Graduation          1708      Together   
1100            3       11     1960.0      Master          1379      Together   
7224            3       11     1977.0  Graduation          1464       Married   

      NKids  NTeens    Income  MntWines  ...  NWebPurchases  \
ID                                       ...                  
5314    0.0     1.0   40689.0       270  ...              7   
2493    0.0     1.0   61286.0       356  ...              7   
4931    0.0     0.0  157146.0         1  ...              0   
1100    1.0     2.0   41275.0        24  ...              3   
7224    NaN     NaN   31353.0        10  ...              1   

      NCatalogPurchases  NStorePurchases  NWebVisitsMonth  Response1  \
ID                                                                     
5314                  1                5                8          0   
2493                  1                8                5          0   
4931                 28                0                1          0   
1100                  1                3                5          0   
7224                  1                2                8          0   

      Response2  Response3  Response4  Response5  Response  
ID                                                          
5314          0          0          0          0         0  
2493          0          0          0          0         0  
4931          0          0          0          0         0  
1100          0          0          0          0         0  
7224          0          0          0          0         0  

[5 rows x 28 columns]

<br>

##### **2.2** Converting BirthYear to Age

In [10]:
# Setup reference date as the first day of 2018
reference_year = "2018"
# Compute number of days since the reference date 
age = 2018 - data["BirthYear"]
# Insert a new column 
data.insert(loc=4, column="Age", value=age)

# Deleting the 'BirthYear' feature, as it is being replaced by Age, that contained the same information
del data["BirthYear"]

# Checking the data
data.head()

CostContact  Revenue   Education   Age  CustomerDays MaritalStatus  \
ID                                                                         
5314            3       11  Graduation  67.0          1750      Together   
2493            3       11  Graduation  52.0          1613       Married   
4931            3       11  Graduation  41.0          1708      Together   
1100            3       11      Master  58.0          1379      Together   
7224            3       11  Graduation  41.0          1464       Married   

      NKids  NTeens    Income  MntWines  ...  NWebPurchases  \
ID                                       ...                  
5314    0.0     1.0   40689.0       270  ...              7   
2493    0.0     1.0   61286.0       356  ...              7   
4931    0.0     0.0  157146.0         1  ...              0   
1100    1.0     2.0   41275.0        24  ...              3   
7224    NaN     NaN   31353.0        10  ...              1   

      NCatalogPurchases  NStorePurchases  NWebVisitsMonth  Response1  \
ID                                                                     
5314                  1                5                8          0   
2493                  1                8                5          0   
4931                 28                0                1          0   
1100                  1                3                5          0   
7224                  1                2                8          0   

      Response2  Response3  Response4  Response5  Response  
ID                                                          
5314          0          0          0          0         0  
2493          0          0          0          0         0  
4931          0          0          0          0         0  
1100          0          0          0          0         0  
7224          0          0          0          0         0  

[5 rows x 28 columns]

<br><br><br>

### **3** Identification and treatment of possible data inconsistencies

#### **3.1** Duplicates

It is important to check if there are repeated rows, that is, duplicates. Having two individuals or more with the same exact information is very unlikely, therefore, when there are duplicates, it is probably due to some mistake and they have to be removed.

Firtly, we use the ``.shape`` method to recall the shape of the dataframe in hands.

In [11]:
data.shape

(2240, 28)

It is applied the method ``.drop_duplicates`` to remove the duplicates from the data. Then, it is relevant to check the dataframe's shape to know how many rows were removed.

In [12]:
data = data.drop_duplicates()
data.shape

(2077, 28)

The shape of the daframe changed after the elimination of duplicates, that is, the amount of rows decreased and, of course, the amount of columns stayed the same. The rows drecreased from 2240 to 2070, which means that 170 rows were eliminated and were, therefore, duplicates.
<br>
<br>

#### **3.2** Missing values

When a dataset presents missing values one can either remove those records or impute. Traditionally, the missing values in the numerical columns are imputed with central tendency measures.

As previously observed, the columns that have missing values are:
* BirthYear, replaced by Age;
* Education;
* NKids;
* NTeens;
* Income.

It is understandable that some of these features can be associated with sensitive topics and therefore some people choose not to share this information. For example, people with the highest and the lowest incomes may decide not to disclose it.

<br>

**3.2.1** To have a better overview of the missing values it is created a dataframe in which the missing values are represented as True and the non-missing values as False.

In [13]:
# Considers empty strings and numpy.inf as missing values
pd.options.mode.use_inf_as_na = True

# It is created a dataframe (data_na) in which the missing values are represented as True and the remaning as False 
data_na = data.isna()

# Visualize data_na DataFrame
display("DataFrame with the missing values as True:", data_na.tail())

'DataFrame with the missing values as True:'

CostContact  Revenue  Education    Age  CustomerDays  MaritalStatus  \
ID                                                                           
3887         False    False      False  False         False          False   
5181         False    False      False  False         False          False   
5758         False    False      False  False         False          False   
9729         False    False      False  False         False          False   
10478        False    False      False  False         False          False   

       NKids  NTeens  Income  MntWines  ...  NWebPurchases  NCatalogPurchases  \
ID                                      ...                                     
3887   False   False   False     False  ...          False              False   
5181   False   False   False     False  ...          False              False   
5758   False   False   False     False  ...          False              False   
9729   False   False   False     False  ...          False              False   
10478  False   False   False     False  ...          False              False   

       NStorePurchases  NWebVisitsMonth  Response1  Response2  Response3  \
ID                                                                         
3887             False            False      False      False      False   
5181             False            False      False      False      False   
5758             False            False      False      False      False   
9729             False            False      False      False      False   
10478            False            False      False      False      False   

       Response4  Response5  Response  
ID                                     
3887       False      False     False  
5181       False      False     False  
5758       False      False     False  
9729       False      False     False  
10478      False      False     False  

[5 rows x 28 columns]

<br>

Using the previous dataframe, we can count the amount of missing values that exist in each column (considering only the columns that have at least one). The boolean values True and False are also defined as the number 1 and 0, respectively. Therefore, the sum of the values in each row (0 if it is False and 1 if True), of a specific column, results in the amount of missing values in that column.

To count those missing values is used the function missing_values(), which performs the respective sum and stores the result in a dataframe in an absolute and relative format.

In [14]:
def missing_values(df, use_inf_as_na=True):       
    pd.options.mode.use_inf_as_na = use_inf_as_na
    # Counts the number of missing values in each row and stores it in na_count
    data_na = data.isna().sum() 
    # Identifies columns with at least one missing values
    features_na = data_na[data_na > 0]    
    # Returns a dataframe with the columns with missing values and corresponding amount and proportion
    data_na = pd.DataFrame.from_dict({"Number of missing values": features_na, 
                                    "Proportion of missing values": features_na.divide(df.shape[0])})
    return data_na

# Using the missing_values 
missing_values(data, True)

Number of missing values  Proportion of missing values
Education                        31                      0.014925
Age                              22                      0.010592
NKids                            28                      0.013481
NTeens                           28                      0.013481
Income                           51                      0.024555

By using this function it is clear that the Income (which has 51 missing values) is the feature with the largest amount of missing values and BirthYear with the least (having 22 missing values) among the features that contain this values.

<br>

**3.2.2** To overcome the problem of missing values, it is used the ``.fillna()`` method. In the code below:
1.  For each numeric column, imputes the missing values with the mean of the respective column;
2.  For each categorical column, imputes the missing values with the mode of the respective column;

In [15]:
# Using the the missing_values function on the dataframe data
display("Before imputing:", missing_values(data))

# Replacing the missing values by the mean for numerical columns
data["Age"].fillna(value=data["Age"].mean(), inplace=True)
data["NKids"].fillna(value=data["NKids"].mean(), inplace=True)
data["NTeens"].fillna(value=data["NTeens"].mean(), inplace=True)
data["Income"].fillna(value=data["Income"].mean(), inplace=True)

# Replacing the missing values by the mode for the categorical column
data["Education"].fillna(value=data["Education"].mode().values[0], inplace=True)

# Calling missing_values function to check if there is still missing values
display("After imputing:", missing_values(data))

'Before imputing:'

Number of missing values  Proportion of missing values
Education                        31                      0.014925
Age                              22                      0.010592
NKids                            28                      0.013481
NTeens                           28                      0.013481
Income                           51                      0.024555

'After imputing:'

Empty DataFrame
Columns: [Number of missing values, Proportion of missing values]
Index: []

The tables are meant to show the amount of missing values and its proportion to the number of rows of the column, as previously displayed. However, these two tables show those same values before and after imputation. Thus, after imputation, we can observe that all the empty spaces were filled, not remaining any missing values.
<br><br>

##### **3.2.3** Transforming float to integer
Following up the treatment of missing values, it is now possible to change the data type of the respective features.

This process is accomplished by using the ``.astype()`` method, which is applied to the three features mentioned (Age, NKids and NTeens).

In [16]:
# Converting the data type of Age to integer
data["Age"] = data['Age'].astype('int')
# Converting the data type of NKids to integer
data["NKids"] = data['NKids'].astype('int')
# Converting the data type of NTeens to integer
data["NTeens"] = data['NTeens'].astype('int')

# Checking the data types
print("Age data type:", data["Age"].dtype,
     "\nNKids data type:", data["NKids"].dtype,
     "\nNTeens data type:", data["NTeens"].dtype)

Age data type: int32 
NKids data type: int32 
NTeens data type: int32


<br><br>

#### **3.3** Extreme/strange/unlikely values

Throughout this notebook, there were some features mentioned due to some corresponding characteristic or specific values that seemed inconsistent, for instance:
* CostContact and Revenue have a null standard deviation and therefore will not be useful to the model.
* The 99999999.0 value of the feature income is extreme, because it is an isolated number, that is, sigmificantly larger than any other value of that feature. 
* The minimum value of BirthYear is 1893, which is a strange number given that the person who inserted it would have 125 years. This feature was converted to Age, and therefore, it is important to look into the values of this new feature and find the unlinke values, such as 125 years.

<br>

##### **3.3.1** Eliminating features with null standard deviation

In [17]:
# Dropping the features "CostContact" and "Revenue"
data = data.drop(["CostContact", "Revenue"], axis=1)

<br>

##### **3.3.2** Eliminating extreme value from income

Considering that the minimum value of Income is 1730 and the maximum is 99999999, it is relevant to check if this last value is actually an isolated case. To accomplish that, we will perform some conditions on the Income feature.

In [18]:
display("Income higher than 200 000:",data[data["Income"]>200000])

'Income higher than 200 000:'

Education  Age  CustomerDays MaritalStatus  NKids  NTeens  \
ID                                                                   
9432    Graduation   41          1674      Together      1       0   
9135  SchoolOfLife   68          1974      Together      1       1   

          Income  MntWines  MntFruits  MntMeatProducts  ...  NWebPurchases  \
ID                                                      ...                  
9432    666666.0         9         14               18  ...              3   
9135  99999999.0        13          2               21  ...              1   

      NCatalogPurchases  NStorePurchases  NWebVisitsMonth  Response1  \
ID                                                                     
9432                  1                3                6          0   
9135                  1                2                8          0   

      Response2  Response3  Response4  Response5  Response  
ID                                                          
9432          0          0          0          0         0  
9135          0          0          0          0         0  

[2 rows x 26 columns]

In the table above we have the rows of the dataframe 'data' in which the Income feature is larger than 200 000. There are two rows, corresponding to the income values 666666.0 and 99999999.0. Those values are both large considering the quartiles of this feature. However, 99999999.0 is significantly larger than 666666.0, which makes it a considerable extreme of this column. Hence, it is benefitial for the model that this value is removed.

To be able to remove rows easily from now on, it is created a function (remove_rows) that makes the process simpler. This function has two parameters, the column and the value. The column corresponds to the feature, from which we intend to remove a specific value. The parameter value is that specific value that is intended to be removed.

In [19]:
def remove_rows(Column, value):
    # Getting the row of the value to be removed 
    index = data.loc[(data[Column] == value), :].index
    # Removing row
    data.drop(index, inplace=True)

In [20]:
# Using the function remove_rows() to remove the row in which Income has the extreme value
remove_rows('Income', 99999999.0)

<br>
To check if the code is working we can build the previous table again.

In [21]:
display("Income higher than 200 000:",data[data["Income"]>200000])

'Income higher than 200 000:'

Education  Age  CustomerDays MaritalStatus  NKids  NTeens    Income  \
ID                                                                           
9432  Graduation   41          1674      Together      1       0  666666.0   

      MntWines  MntFruits  MntMeatProducts  ...  NWebPurchases  \
ID                                          ...                  
9432         9         14               18  ...              3   

      NCatalogPurchases  NStorePurchases  NWebVisitsMonth  Response1  \
ID                                                                     
9432                  1                3                6          0   

      Response2  Response3  Response4  Response5  Response  
ID                                                          
9432          0          0          0          0         0  

[1 rows x 26 columns]

It is clear that the extreme value mentioned was eliminated.

<br>

##### **3.3.3** Eliminating unlikely values in feature Age

Firstly, we can check the minimum and maximum values of Age to know in which way it is necessary to restrict this feature.

In [22]:
display("Age min:",data["Age"].min(), "Age max:", data["Age"].max())

'Age min:'

22

'Age max:'

125

The minimum value of Age is 22 which does not cause any problems. However, its maximum, as already mentioned, it is 125, which is impossible. 

Although possible, it is very unlikely for a person to live up to 100 years old, therefore, we will eliminate the rows in which the value for Age is higher than 100. Thus, we are considering in the model people that are 100 years old or younger.

In the code below we check the rows where Age is higher than 100.

In [23]:
display("Age higher than 100:",data[data["Age"]>100])

'Age higher than 100:'

Education  Age  CustomerDays MaritalStatus  NKids  NTeens   Income  \
ID                                                                         
11004  2n Cycle  125          1325        Single      0       1  60182.0   
7829   2n Cycle  118          1558      Divorced      1       0  36640.0   
1150        PhD  119          1558      Together      0       0  83532.0   

       MntWines  MntFruits  MntMeatProducts  ...  NWebPurchases  \
ID                                           ...                  
11004         8          0                5  ...              1   
7829         15          6                8  ...              2   
1150        755        144              562  ...              4   

       NCatalogPurchases  NStorePurchases  NWebVisitsMonth  Response1  \
ID                                                                      
11004                  0                2                4          0   
7829                   1                2                5          0   
1150                   6                4                1          0   

       Response2  Response3  Response4  Response5  Response  
ID                                                           
11004          0          0          0          0         0  
7829           0          0          0          0         0  
1150           0          0          0          1         0  

[3 rows x 26 columns]

It is observed in the table above that there are only three rows in the dataset in which age is higher than 100. Knowing that the oldest person ever alive was Jiroemon Kimura, who lived to the age of 116, it is not possible for the indiduals corresponding to those three rows to be alive. Therefore, they will be removed from the data.

In [24]:
# Using the remove_rows() function to remove the observed rows in which Age has a value greater than 100
remove_rows('Age', 118)
remove_rows('Age', 119)
remove_rows('Age', 125)

<br>
To check if the code is working we can build the previous table again.

In [25]:
display("Age higher than 100:",data[data["Age"]>100])

'Age higher than 100:'

Empty DataFrame
Columns: [Education, Age, CustomerDays, MaritalStatus, NKids, NTeens, Income, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, Complain, Recency, NDealsPurchases, NWebPurchases, NCatalogPurchases, NStorePurchases, NWebVisitsMonth, Response1, Response2, Response3, Response4, Response5, Response]
Index: []

[0 rows x 26 columns]

The table is now empty, therefore there are not values of the column Age greater than 100. Hence, the unlikely values of Age were successfully removed.
<br><br><br>

##### **3.3.4** Eliminating strange values in the MaritalStatus feature
We can now look into the feature MaritalStatus, that was previously mentioned due to its strange number of unique values.

First, it is verified the unique values of the column, using the ``.unique()`` method.

In [26]:
data["MaritalStatus"].unique()

array(['Together', 'Married', 'Single', 'Divorced', 'Widow', 'Alone',
       'YOLO', 'Absurd'], dtype=object)

By visualizing this array, it is clear that there are some values which are pointless, for instance:
* The value 'Alone' is equivalent to the value 'Single', and therefore, it is important to replace it by this second value.
* 'YOLO' and 'Absurd' are not meaningful for the model. Thus, this values will be considered to be removed.


<br>


**3.3.4.1** Replacing 'Alone' and 'Together' by 'Single'

To acknowledge which rows have the value 'Alone' or 'Single' in the MaritalStatus column, it is relevant to visualize the table containg only the rows with those values. The code below displays that table.

In [27]:
data.loc[(data["MaritalStatus"] == "Alone") | (data["MaritalStatus"] == "Together"), :]

Education  Age  CustomerDays MaritalStatus  NKids  NTeens    Income  \
ID                                                                           
5314  Graduation   67          1750      Together      0       1   40689.0   
4931  Graduation   41          1708      Together      0       0  157146.0   
1100      Master   58          1379      Together      1       2   41275.0   
3537  Graduation   37          1963      Together      1       0   31089.0   
807       Master   49          1416      Together      1       1   44602.0   
...          ...  ...           ...           ...    ...     ...       ...   
2445         PhD   44          1639      Together      1       0   28440.0   
5782    2n Cycle   36          1556      Together      0       1   58582.0   
8213    2n Cycle   38          1879      Together      1       0   14515.0   
7873         PhD   45          1640      Together      1       1   63516.0   
6424  Graduation   41          1432      Together      1       0   33178.0   

      MntWines  MntFruits  MntMeatProducts  ...  NWebPurchases  \
ID                                          ...                  
5314       270          3               27  ...              7   
4931         1          0             1725  ...              0   
1100        24          4               22  ...              3   
3537        31          3               31  ...              3   
807        167          2               89  ...              6   
...        ...        ...              ...  ...            ...   
2445        38          1               24  ...              2   
5782       402         35              106  ...              9   
8213         6          4                9  ...              2   
7873       141         11              114  ...              4   
6424        12          6                2  ...              1   

      NCatalogPurchases  NStorePurchases  NWebVisitsMonth  Response1  \
ID                                                                     
5314                  1                5                8          0   
4931                 28                0                1          0   
1100                  1                3                5          0   
3537                  0                4                8          0   
807                   1                4                8          0   
...                 ...              ...              ...        ...   
2445                  0                3                8          0   
5782                  2                9                5          0   
8213                  2                3                7          0   
7873                  1                7                5          0   
6424                  0                3                4          0   

      Response2  Response3  Response4  Response5  Response  
ID                                                          
5314          0          0          0          0         0  
4931          0          0          0          0         0  
1100          0          0          0          0         0  
3537          0          0          0          0         0  
807           0          0          0          0         0  
...         ...        ...        ...        ...       ...  
2445          0          0          0          0         0  
5782          0          0          1          0         0  
8213          0          0          0          0         1  
7873          0          0          0          0         0  
6424          0          0          0          0         0  

[526 rows x 26 columns]

It is possible to observe that there are only three rows that satisfy the condition aforementioned. After replacing those values successfully, there will be none.

To achieve that value replacement, it is used the  ``.replace()`` method, which will replace every value in the dataframe corresponding to 'Alone' and 'Together' by 'Single'.

In [28]:
data = data.replace('Alone', 'Single')
data = data.replace('Together', 'Single')

To guarantee that the code was successful, it is used the above-entered code to display all the rows in which the feature MaritalStatus has the value 'Alone'.

In [29]:
data.loc[(data["MaritalStatus"] == "Alone"), :]

Empty DataFrame
Columns: [Education, Age, CustomerDays, MaritalStatus, NKids, NTeens, Income, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, Complain, Recency, NDealsPurchases, NWebPurchases, NCatalogPurchases, NStorePurchases, NWebVisitsMonth, Response1, Response2, Response3, Response4, Response5, Response]
Index: []

[0 rows x 26 columns]

The table is empty, therefore, there is no longer a redundant value 'Alone' in the dataset 'data'.


<br>

**3.3.4.2** Removing 'YOLO' and 'Absurd'

To acknowledge which rows have the value 'YOLO' and 'Absurd' in the MaritalStatus column, it is relevant to visualize the table containg only the rows with those values. The code below displays that table.

In [30]:
data.loc[(data["MaritalStatus"] == "YOLO") | (data["MaritalStatus"] == "Absurd"), :]

Education  Age  CustomerDays MaritalStatus  NKids  NTeens   Income  \
ID                                                                           
492           PhD   45          1901          YOLO      0       1  48432.0   
7734   Graduation   25          1839        Absurd      0       0  79244.0   
11133         PhD   45          1901          YOLO      0       1  48432.0   
4369       Master   61          1452        Absurd      0       0  65487.0   

       MntWines  MntFruits  MntMeatProducts  ...  NWebPurchases  \
ID                                           ...                  
492         322          3               50  ...              7   
7734        471        102              125  ...              4   
11133       322          3               50  ...              7   
4369        240         67              500  ...              3   

       NCatalogPurchases  NStorePurchases  NWebVisitsMonth  Response1  \
ID                                                                      
492                    1                6                8          0   
7734                  10                7                1          1   
11133                  1                6                8          0   
4369                   5                6                2          0   

       Response2  Response3  Response4  Response5  Response  
ID                                                           
492            0          0          0          0         0  
7734           0          0          0          1         1  
11133          0          0          0          0         1  
4369           0          0          0          0         0  

[4 rows x 26 columns]

By visualizing the table, it is observed that there are only four rows that contain the values above-mentioned, which is a small percentage of the current 2073 rows . Hence, it is not detrimental for the model to remove this meaningless values.

To remove this values from the dataset, it is used the previously created function ``remove_rows()``.

In [31]:
remove_rows('MaritalStatus', 'YOLO')
remove_rows('MaritalStatus', 'Absurd')

To check if the code was successful, it is used the above-entered code to display all the rows in which the feature MaritalStatus presents the values 'YOLO' and 'Absurd'.

In [32]:
data.loc[(data["MaritalStatus"] == "YOLO") | (data["MaritalStatus"] == "Absurd"), :]

Empty DataFrame
Columns: [Education, Age, CustomerDays, MaritalStatus, NKids, NTeens, Income, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, Complain, Recency, NDealsPurchases, NWebPurchases, NCatalogPurchases, NStorePurchases, NWebVisitsMonth, Response1, Response2, Response3, Response4, Response5, Response]
Index: []

[0 rows x 26 columns]

The dataframe is empty, which implies that the values mentioned were effectively removed.

<br><br><br>

### **4.** Identification and treatment of inconsistent facts

Over the process of studying and transforming the features of the 'data' dataset, it was possible to recognize some facts that need to be satisfied among some features. For instance: 

* For an individual to have a value different from 0 in Revenue, it would have to respond to the campaign. Hence, if the feature Response is 0, the Revenue would have to be 0 as well, and if it the Response is 1, the Revenue is 11 (which corresponds to the monetary units received by each customer that positively responded to the campaign). Given the fact that every single value of Revenue is 11, the solution would be to replace the value 11 by 0, when Response was equal to 0. However, it was observed that the feature Revenue had a null standard deviation and therefore, was not useful for the model. If that solution was actually put in practise, Revenue would provide exactly the same information as Response, which would not be helpful. Thus, no change is needed in regards to the independent variable mentioned.
* The MntGoldProds variable should not be larger than the amount of money spent among all the products (that is, the sum of the variables MntFishProducts, MntMeatProducts, MntFruits, MntSweetProducts, MntWines).
* If an individual has never purchased a product from the company (NWebPurchases, NCatalogPurchases and NStorePurchases are equal to 0), then it does not make sense for Recency to have a numerical value. Instead, it should be filled with a 'Not Applicable' value. 
* If NWebPurchases, NCatalogPurchases and NStorePurchases are equal to 0, it is conflicting to have values larger than 0 on the amount spent in products (MntMeatProducts, MntWines, MntFruits, MntFishProducts, MntSweetProducts)). Therefore, in those cases, the values of the last five features mentioned, that are larger than 0, should be replaced by 0.
* The NDealsPurchases should be null when no purchases are made, that is, when NWebPurchases, NCatalogPurchases and NStorePurchases are equal to 0. This is because the purchases represented by NDealsPurchases should be made in one of the three possible ways, that is, through the store, the website or a catalog.

<br>


##### **4.1** MntGoldProds feature should not be larger than the amount of money spent among all the products

In order to know if this fact is always satisfied, the rows of the dataset in which that is not the case are selected. Hence, in the code below, it is made the condition of the feature MntGoldProds being larger than the total amount of money spent among all products, within the data, so that we are able get the desirable rows. The ``.T`` method is also used, to allow us to check the respective features.

In [33]:
data[data['MntGoldProds'] > data["MntFishProducts"] + data['MntMeatProducts'] + data["MntFruits"] + data['MntSweetProducts'] +  data['MntWines']].T

ID                        5255    6237        10311   4246 
Education            Graduation     PhD  Graduation  Master
Age                          32      52          49      36
CustomerDays               1776    1486        1549    1481
MaritalStatus            Single  Single     Married  Single
NKids                         1       0           0       0
NTeens                        0       2           1       0
Income             101658.71619  7144.0      4428.0  6560.0
MntWines                      5      81          16      67
MntFruits                     1       4           4      11
MntMeatProducts               3      33          12      26
MntFishProducts               3       5           2       4
MntSweetProducts            263       2           4       3
MntGoldProds                362     291         321     262
Complain                      0       0           0       0
Recency                      19      92           0       2
NDealsPurchases               0       0           0       0
NWebPurchases                27      23          25       1
NCatalogPurchases             0       1           0       0
NStorePurchases               0       1           0       1
NWebVisitsMonth               1       0           1      17
Response1                     0       0           0       0
Response2                     0       0           0       0
Response3                     0       0           0       0
Response4                     0       0           0       0
Response5                     0       0           0       0
Response                      0       0           0       0

In the table above, it is observed that there are four individuals for whom this condition is True. To solve this violation of the fact, it is removed the small amount of rows in which that happens, using the function ``remove_rows()``.

In [34]:
remove_rows('MntGoldProds', 262)
remove_rows('MntGoldProds', 362)
remove_rows('MntGoldProds', 291)
remove_rows('MntGoldProds', 321)

To check if the rows were effectively removed, we can apply the above-entered condition on the data anew.

In [35]:
data[data['MntGoldProds'] > data["MntFishProducts"] + data['MntMeatProducts'] + data["MntFruits"] + data['MntSweetProducts'] +  data['MntWines']]

Empty DataFrame
Columns: [Education, Age, CustomerDays, MaritalStatus, NKids, NTeens, Income, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, Complain, Recency, NDealsPurchases, NWebPurchases, NCatalogPurchases, NStorePurchases, NWebVisitsMonth, Response1, Response2, Response3, Response4, Response5, Response]
Index: []

[0 rows x 26 columns]

By observing the empty table, it is evident that there are no longer indivuals whose information contradicts the respective fact.

<br>

##### **4.2** If an individual has never purchased a product from the company, then Recency is 'Not Applicable'

First, it is relevant to verify if there are individuals that do not respect this fact. To do that, it is applied a condition in the data in which the purchase values in stores, catalogs and the website are all equal to 0. This is perfomed in the code below, and it is, once again, employed the method ``.T`` to transpose the resulting dataframe and be able to visualize every column.

In [36]:
data[(data["NCatalogPurchases"] == 0) & (data["NStorePurchases"] == 0) & (data["NWebPurchases"] == 0)].T

ID                      5555     9931        11110     11181       3955   \
Education          Graduation      PhD  Graduation       PhD  Graduation   
Age                        43       55          45        69          53   
CustomerDays             1424     1288        1724      1586        1289   
MaritalStatus        Divorced  Married      Single   Married    Divorced   
NKids                       0        1           1         0           0   
NTeens                      0        1           0         0           0   
Income               153924.0   4023.0      3502.0  156924.0      4861.0   
MntWines                    1        5           2         2           2   
MntFruits                   1        0           1         1           1   
MntMeatProducts             1        1           1         2           1   
MntFishProducts             1        1           0         1           1   
MntSweetProducts            1        1           0         1           0   
MntGoldProds                1        1           1         1           1   
Complain                    0        0           0         0           0   
Recency                    81       29          56        85          20   
NDealsPurchases             0       15           0         0           0   
NWebPurchases               0        0           0         0           0   
NCatalogPurchases           0        0           0         0           0   
NStorePurchases             0        0           0         0           0   
NWebVisitsMonth             0       19          14         0          14   
Response1                   0        0           0         0           0   
Response2                   0        0           0         0           0   
Response3                   0        0           0         0           0   
Response4                   0        0           0         0           0   
Response5                   0        0           0         0           0   
Response                    0        0           0         0           0   

ID                      6862   
Education          Graduation  
Age                        47  
CustomerDays             1324  
MaritalStatus        Divorced  
NKids                       0  
NTeens                      0  
Income                 1730.0  
MntWines                    1  
MntFruits                   1  
MntMeatProducts             3  
MntFishProducts             1  
MntSweetProducts            1  
MntGoldProds                1  
Complain                    0  
Recency                    65  
NDealsPurchases            15  
NWebPurchases               0  
NCatalogPurchases           0  
NStorePurchases             0  
NWebVisitsMonth            20  
Response1                   0  
Response2                   0  
Response3                   0  
Response4                   0  
Response5                   0  
Response                    0

The dataframe above enables the acknowledgement that there are six individuals that do not statisfy the fact established. To solve that issue, the values of the Recency feature are replaced by 'Not Applicable'. 

In the code below, we start by creating a function to replace specific values in the dataframe, so that it is possible to do that more easily from now on.

In [37]:
def replace_values(Column, indexes, replace_value):
    # Replacing values of a certain column by the 'replace_value'
    data.loc[indexes, Column] = replace_value

It is now possible to replace the respective values in just one cell.

In [38]:
replace_values('Recency', [3955, 5555, 6862, 9931, 11110, 11181],'Not Applicable' )

Again, it is relevant to check if the process was successful by running the condition one more time.

In [39]:
data[(data["NCatalogPurchases"] == 0) & (data["NStorePurchases"] == 0) & (data["NWebPurchases"] == 0)].T

ID                          5555            9931            11110  \
Education              Graduation             PhD      Graduation   
Age                            43              55              45   
CustomerDays                 1424            1288            1724   
MaritalStatus            Divorced         Married          Single   
NKids                           0               1               1   
NTeens                          0               1               0   
Income                   153924.0          4023.0          3502.0   
MntWines                        1               5               2   
MntFruits                       1               0               1   
MntMeatProducts                 1               1               1   
MntFishProducts                 1               1               0   
MntSweetProducts                1               1               0   
MntGoldProds                    1               1               1   
Complain                        0               0               0   
Recency            Not Applicable  Not Applicable  Not Applicable   
NDealsPurchases                 0              15               0   
NWebPurchases                   0               0               0   
NCatalogPurchases               0               0               0   
NStorePurchases                 0               0               0   
NWebVisitsMonth                 0              19              14   
Response1                       0               0               0   
Response2                       0               0               0   
Response3                       0               0               0   
Response4                       0               0               0   
Response5                       0               0               0   
Response                        0               0               0   

ID                          11181           3955            6862   
Education                     PhD      Graduation      Graduation  
Age                            69              53              47  
CustomerDays                 1586            1289            1324  
MaritalStatus             Married        Divorced        Divorced  
NKids                           0               0               0  
NTeens                          0               0               0  
Income                   156924.0          4861.0          1730.0  
MntWines                        2               2               1  
MntFruits                       1               1               1  
MntMeatProducts                 2               1               3  
MntFishProducts                 1               1               1  
MntSweetProducts                1               0               1  
MntGoldProds                    1               1               1  
Complain                        0               0               0  
Recency            Not Applicable  Not Applicable  Not Applicable  
NDealsPurchases                 0               0              15  
NWebPurchases                   0               0               0  
NCatalogPurchases               0               0               0  
NStorePurchases                 0               0               0  
NWebVisitsMonth                 0              14              20  
Response1                       0               0               0  
Response2                       0               0               0  
Response3                       0               0               0  
Response4                       0               0               0  
Response5                       0               0               0  
Response                        0               0               0

The resulting dataframe shows that the desirable fact is now being fulfilled.

<br>

##### **4.3** If NWebPurchases, NCatalogPurchases and NStorePurchases are equal to 0, then the amount spent in products is 0.

To know whether this fact is being satisfied, it is used the above-entered condition in the code below. This happens, since this two facts are based on the value of NWebPurchases, NCatalogPurchases and NStorePurchases being equal to 0.

In [40]:
data[(data["NWebPurchases"] == 0) & (data["NCatalogPurchases"] == 0) & (data["NStorePurchases"] == 0)].T

ID                          5555            9931            11110  \
Education              Graduation             PhD      Graduation   
Age                            43              55              45   
CustomerDays                 1424            1288            1724   
MaritalStatus            Divorced         Married          Single   
NKids                           0               1               1   
NTeens                          0               1               0   
Income                   153924.0          4023.0          3502.0   
MntWines                        1               5               2   
MntFruits                       1               0               1   
MntMeatProducts                 1               1               1   
MntFishProducts                 1               1               0   
MntSweetProducts                1               1               0   
MntGoldProds                    1               1               1   
Complain                        0               0               0   
Recency            Not Applicable  Not Applicable  Not Applicable   
NDealsPurchases                 0              15               0   
NWebPurchases                   0               0               0   
NCatalogPurchases               0               0               0   
NStorePurchases                 0               0               0   
NWebVisitsMonth                 0              19              14   
Response1                       0               0               0   
Response2                       0               0               0   
Response3                       0               0               0   
Response4                       0               0               0   
Response5                       0               0               0   
Response                        0               0               0   

ID                          11181           3955            6862   
Education                     PhD      Graduation      Graduation  
Age                            69              53              47  
CustomerDays                 1586            1289            1324  
MaritalStatus             Married        Divorced        Divorced  
NKids                           0               0               0  
NTeens                          0               0               0  
Income                   156924.0          4861.0          1730.0  
MntWines                        2               2               1  
MntFruits                       1               1               1  
MntMeatProducts                 2               1               3  
MntFishProducts                 1               1               1  
MntSweetProducts                1               0               1  
MntGoldProds                    1               1               1  
Complain                        0               0               0  
Recency            Not Applicable  Not Applicable  Not Applicable  
NDealsPurchases                 0               0              15  
NWebPurchases                   0               0               0  
NCatalogPurchases               0               0               0  
NStorePurchases                 0               0               0  
NWebVisitsMonth                 0              14              20  
Response1                       0               0               0  
Response2                       0               0               0  
Response3                       0               0               0  
Response4                       0               0               0  
Response5                       0               0               0  
Response                        0               0               0

In the dataframe above, it is noticeable that the fact is not being fulfilled. That is, the values of the amounts spent on products (MntMeatProducts, MntWines, MntFruits,MntFishProducts, MntSweetProducts) are larger than 0 while the mentioned purchase values are null. In this case, it is also possible to solve the situation by replacing the values that do not follow the desirable path. Those values (in MntMeatProducts, MntWines, MntFruits,MntFishProducts, MntSweetProducts)) will be replaced by 0, which is performed in the code below, using the function ``replace_values()``.

In [41]:
replace_values(['MntMeatProducts','MntWines', 'MntFruits','MntFishProducts', 'MntSweetProducts'],[3955, 5555, 6862, 9931, 11110, 11181], 0 )

The previous code is reused to check if the changes were correctly made.

In [42]:
data[(data["NWebPurchases"] == 0) & (data["NCatalogPurchases"] == 0) & (data["NStorePurchases"] == 0)].T

ID                          5555            9931            11110  \
Education              Graduation             PhD      Graduation   
Age                            43              55              45   
CustomerDays                 1424            1288            1724   
MaritalStatus            Divorced         Married          Single   
NKids                           0               1               1   
NTeens                          0               1               0   
Income                   153924.0          4023.0          3502.0   
MntWines                        0               0               0   
MntFruits                       0               0               0   
MntMeatProducts                 0               0               0   
MntFishProducts                 0               0               0   
MntSweetProducts                0               0               0   
MntGoldProds                    1               1               1   
Complain                        0               0               0   
Recency            Not Applicable  Not Applicable  Not Applicable   
NDealsPurchases                 0              15               0   
NWebPurchases                   0               0               0   
NCatalogPurchases               0               0               0   
NStorePurchases                 0               0               0   
NWebVisitsMonth                 0              19              14   
Response1                       0               0               0   
Response2                       0               0               0   
Response3                       0               0               0   
Response4                       0               0               0   
Response5                       0               0               0   
Response                        0               0               0   

ID                          11181           3955            6862   
Education                     PhD      Graduation      Graduation  
Age                            69              53              47  
CustomerDays                 1586            1289            1324  
MaritalStatus             Married        Divorced        Divorced  
NKids                           0               0               0  
NTeens                          0               0               0  
Income                   156924.0          4861.0          1730.0  
MntWines                        0               0               0  
MntFruits                       0               0               0  
MntMeatProducts                 0               0               0  
MntFishProducts                 0               0               0  
MntSweetProducts                0               0               0  
MntGoldProds                    1               1               1  
Complain                        0               0               0  
Recency            Not Applicable  Not Applicable  Not Applicable  
NDealsPurchases                 0               0              15  
NWebPurchases                   0               0               0  
NCatalogPurchases               0               0               0  
NStorePurchases                 0               0               0  
NWebVisitsMonth                 0              14              20  
Response1                       0               0               0  
Response2                       0               0               0  
Response3                       0               0               0  
Response4                       0               0               0  
Response5                       0               0               0  
Response                        0               0               0

In the dataframe above, the features corresponding to the amount of money spent in product are filled with null values, therefore, the desired fact is now being statisfied. 

<br>

##### **4.4** The NDealsPurchases should be null when no purchases are made

Once again it is used the condition of the purchase features being all null. Hence, the above-entered code will be reused above with an additional '&' within the condition to establish that the NDealsPurchases feature is larger than 0.

In [43]:
data[(data["NCatalogPurchases"] == 0) & (data["NStorePurchases"] == 0) & (data["NWebPurchases"] == 0) & (data["NDealsPurchases"] > 0)].T

ID                           9931            6862
Education                     PhD      Graduation
Age                            55              47
CustomerDays                 1288            1324
MaritalStatus             Married        Divorced
NKids                           1               0
NTeens                          1               0
Income                     4023.0          1730.0
MntWines                        0               0
MntFruits                       0               0
MntMeatProducts                 0               0
MntFishProducts                 0               0
MntSweetProducts                0               0
MntGoldProds                    1               1
Complain                        0               0
Recency            Not Applicable  Not Applicable
NDealsPurchases                15              15
NWebPurchases                   0               0
NCatalogPurchases               0               0
NStorePurchases                 0               0
NWebVisitsMonth                19              20
Response1                       0               0
Response2                       0               0
Response3                       0               0
Response4                       0               0
Response5                       0               0
Response                        0               0

In the dataframe above, it is observed that there are two individuals that do not respect the respective fact. To solve this situation, the value of the NDealsPurchases feature for those two individuals will be replaced by 0. This will be performed alongside the usage of the function ``replace_values()``, as displayed in the code below.

In [44]:
replace_values('NDealsPurchases', [9931, 6862], 0)

In the code below, the previous condition is executed again, to check if the replacement was sucessful.

In [45]:
data[(data["NCatalogPurchases"] == 0) & (data["NStorePurchases"] == 0) & (data["NWebPurchases"] == 0) & (data["NDealsPurchases"] > 0)]

Empty DataFrame
Columns: [Education, Age, CustomerDays, MaritalStatus, NKids, NTeens, Income, MntWines, MntFruits, MntMeatProducts, MntFishProducts, MntSweetProducts, MntGoldProds, Complain, Recency, NDealsPurchases, NWebPurchases, NCatalogPurchases, NStorePurchases, NWebVisitsMonth, Response1, Response2, Response3, Response4, Response5, Response]
Index: []

[0 rows x 26 columns]

The dataframe above is empty, meaning that there is no longer values that conflict with the mentioned fact.

<br><br>

### **5.** Exporting the dataset

Now that the preprocess phase is complete, with the data understood, transformed and cleaned, it is time to start the next one. For us to proceed to the next phase with the dataset developed in this stage, it is necessary to export it. 

In the code below, the dataset is being exported in an excel format with the name "after_preprocess.xlsx". 

In [46]:
data.to_excel("after_preprocess.xlsx")